### PR TITLE
Add staff support for impersonating users

### PR DESCRIPTION
**Pull request by axbarsan**
Closes https://github.com/giantswarm/roadmap/issues/564

This PR adds a new `Impersonation` section to the `Experiments` page.
With it we can act as other users, test permissions or debug problems.

This is only visible to Giant Swarm staff, for now. We can create better UI support for customers later on.

Impersonation is also persisted between sessions. If you get stuck, you can delete the `localStorage` entry from the DevTools Inspector.

[Read more about Kubernetes API impersonation](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation)

## Preview

image

### CODE REVIEW

File: src/lib/OAuth2/OAuth2.ts [hunk 1]
```diff
@@ -40,12 +40,21 @@ export interface IOAuth2SigningKey {
   e: string;
 }
 
+export interface IOAuth2ImpersonationMetadata {
```
Comment (axbarsan) [inline src/lib/OAuth2/OAuth2.ts:43]: Luckily impersonation is also a OAuth2 thing, so we can add it here 🥳

File: src/model/services/mapi/generic/ensureClientAuth.ts [hunk 2]
```diff
@@ -24,6 +24,15 @@ export async function ensureClientAuth(
 
   client.setAuthorizationToken(user.authorizationType, user.idToken);
 
+  const impersonationMetadata = await auth.getImpersonationMetadata();
```
Comment (axbarsan) [inline src/model/services/mapi/generic/ensureClientAuth.ts:27]: Decorating all MAPI requests with the impersonation configuration (if it's set)

**Comment by marians** [comment]
Giving it a test drive now

**Comment by marians** [comment]
A conceptual thing: I think it should it be possible to set more than one group name. SSO users can be member of multiple groups, too. So for testing it would help to be able to apply that. I see that the storage object stores a groups array. So all we have to do I guess is allow input of multiple names. We could simply use white space as separator. Plus add the according info next to the input field.

---

My first test was to create a rolebinding for the `read-all` group in an organization (`giantswarm` on `gollum`). Like this:

```nohighlight
$ kubectl -n org-giantswarm create rolebinding marian-read-only-test --role read-all --group read-only-users
rolebinding.rbac.authorization.k8s.io/marian-read-only-test created
```

Then I used the impersonation form with these settings:

- Username: `username-does-not-matter`
- Group: `read-only-users`

I hit the `Set` button and navigated to the home page (Cluster). There I got a lot of error messages saying something like "user 'username-does-not-matter' is not allowed to get resource XYZ". Then the page disappeared and the Insufficient permissions screen appeared instead.

image

When going through "Try logging in again" I'm running into the same "Insufficient permissions" screen again after logging in. So I was locked out.

I guess this is more or less all expected, because the user is in fact lacking permissions required for happa.

Can we improve the "Insufficient permissions" screen for this case? I think we could:

- Show the impersonated user name and groups instead of the one from the ID token
- If impersonation is in place, tell the user how to  quit impersonation, or offer a button to quit impersonation simply.

---

Question: what are the exact permissions again happa needs for every user so they don't see the "Insufficient permissions" screen?

File: src/components/MAPI/Experiments.tsx [hunk 3]
```diff
@@ -80,6 +122,61 @@ const Experiments: React.FC<IExperimentsProps> = () => {
...
+                label='Username'
+                id='username'
+                name='username'
+                help='The user defined in the RoleBinding.'
+                required={true}
+                value={impersonationUser}
+                onChange={(e) => setImpersonationUser(e.target.value)}
+              />
+              <TextInput
+                label='Group'
```
Comment (marians) [inline src/components/MAPI/Experiments.tsx:163]:
```suggestion
                label='Groups'
```

File: src/components/MAPI/Experiments.tsx [hunk 4]
```diff
@@ -28,6 +32,44 @@ const Experiments: React.FC<IExperimentsProps> = () => {
...
+    new FlashMessage(
+      'Impersonation configured successfully.',
+      messageType.SUCCESS,
+      messageTTL.MEDIUM
+    );
+  };
+
+  const handleClearImpersonation = () => {
+    auth.setImpersonationMetadata(null);
+
```
Comment (kuosandys) [inline src/components/MAPI/Experiments.tsx:65]: Could we clear the form in the UI as well? 🙂

**Comment by axbarsan** [comment]
> A conceptual thing: I think it should it be possible to set more than one group name. SSO users can be member of multiple groups, too. So for testing it would help to be able to apply that. I see that the storage object stores a groups array. So all we have to do I guess is allow input of multiple names. We could simply use white space as separator. Plus add the according info next to the input field.

Yes, the K8s API supports multiple groups, but it requires multiple headers with the same name (`Impersonate-Group`). This is not possible at the moment in our HTTP Client, because internally it uses a map for handling headers. Supporting multiple groups is an improvement and can be added later. A single group is fine for now.

I'll take a look at improving the `Insufficient permissions` screen

File: src/components/MAPI/Experiments.tsx (hunk 4 above)
Comment (axbarsan) [reply src/components/MAPI/Experiments.tsx:65]: Nah that would be too smart 😆

**Comment by axbarsan** [comment]
If impersonating an user, the `Insufficient permissions` page now looks like this:

image

**Review by kuosandys — APPROVED** [review]

**Comment by marians** [comment]
> the K8s API supports multiple groups, but it requires multiple headers with the same name

Oh wow!

**Pull request merged**